tests/php/FilesystemTest.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+namespace SilverStripe\Assets\Tests;
+
+use InvalidArgumentException;
+use SilverStripe\Assets\Filesystem;
+use SilverStripe\Dev\SapphireTest;
+
+class FilesystemTest extends SapphireTest
+{
+    /**
+     * Test paths are joined
+     *
+     * @dataProvider providerTestJoinPaths
+     * @param array $args Arguments to pass to Filesystem::joinPath()
+     * @param string $expected Expected path
+     */
+    public function testJoinPaths($args, $expected)
+    {
+        $joined = Filesystem::joinPaths($args);
+        $this->assertEquals($expected,  $joined);
+    }
+
+    /**
+     * List of tests for testJoinPaths
+     *
+     * @return array
+     */
+    public function providerTestJoinPaths()
+    {
+        $tests = [
+            // Single arg
+            [['/'], '/'],
+            [['\\'], '/'],
+            [['base'], 'base'],
+            [['c:/base\\'], 'c:/base'],
+            // Windows paths
+            [['c:/', 'bob'], 'c:/bob'],
+            [['c:/', '\\bob/'], 'c:/bob'],
+            [['c:\\basedir', '/bob\\'], 'c:/basedir/bob'],
+            // Empty-ish paths to clear out
+            [['/root/dir', '/', ' ', 'next/', '\\'], '/root/dir/next'],
+            [['/', '/', ' ', '/', '\\'], '/'],
+            [['/', '', '',], '/'],
+            [['/root', '/', ' ', '/', '\\'], '/root'],
+            [['', '/root', '/', ' ', '/', '\\'], '/root'],
+            [['', 'root', '/', ' ', '/', '\\'], 'root'],
+            [['\\', '', '/root', '/', ' ', '/', '\\'], '/root'],
+            // join blocks of paths
+            [['/root/dir', 'another/path\\to/join'], '/root/dir/another/path/to/join'],
+        ];
+
+        // Rewrite tests for other filesystems (output arg only)
+        if (DIRECTORY_SEPARATOR !== '/') {
+            foreach ($tests as $index => $test) {
+                $tests[$index][1] = str_replace('/', DIRECTORY_SEPARATOR, $tests[$index][1]);
+            }
+        }
+        return $tests;
+    }
+
+    /**
+     * Test that joinPaths give the appropriate error
+     *
+     * @dataProvider providerTestJoinPathsErrors
+     * @param array $args Arguments to pass to Filesystem::joinPath()
+     * @param string $error Expected path
+     */
+    public function testJoinPathsErrors($args, $error)
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage($error);
+        Filesystem::joinPaths($args);
+    }
+
+    public function providerTestJoinPathsErrors()
```

Comment:
only `test` methods should be public, providers should be `protected`

Reply:
That’s not true in phpunit 5, though I wish it was :-D

Reply:
phpunit docs declare public providers still.

https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers

Can you show me a reference to protected providers being preferred?

Reply:
```md
A data provider method must be public and either return an array of arrays or an object that implements the Iterator interface and yields an array for each iteration step. For each array that is part of the collection the test method will be called with the contents of the array as its arguments.
```

Reply:
Same docs in 7.0 of phpunit.

Reply:
Yikes. I've always made them protected. Maybe it's a Codeception thing, then... My bad